x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java

```diff
@@ -365,7 +364,13 @@ public void clusterStateProcessed(ClusterState oldState, ClusterState newState)
             }
         }
         assert assertNoOtherActiveAutoFollower(newAutoFollowers);
-        this.autoFollowers = autoFollowersCopy.copyAndPutAll(newAutoFollowers).copyAndRemoveAll(removedRemoteClusters);
+
+        Map<String, AutoFollower> updatedFollowers = new HashMap<>(currentAutoFollowers);
+        updatedFollowers.putAll(newAutoFollowers);
+        for (String removedCluster : removedRemoteClusters) {
+            updatedFollowers.remove(removedCluster);
+        }
```

Comment:
> The set supports element removal, which removes the corresponding mapping from the map

Huh, I didn't know that about `keySet()`, nice.

Reply:
IntelliJ says that `removeAll()` is likely to be slow and instead suggests `removedRemoteClusters.forEach(updatedFollowers.keySet()::remove);`

Reply:
Oh, I didn't know about this. This makes me wonder what removeAll is doing under the hood so that it's slower than this naive implementation.

Reply:
https://stackoverflow.com/questions/28671903/the-hashsett-removeall-method-is-surprisingly-slow

Looks like it's for the specific case where the collection passed to remove is a List, and the List is larger than the set being modified; `removeAll()` ends up calling `collection.contains()` on the List, which is a linear scan.  So it probably wouldn't bite us here (the list of clusters to remove will always be smaller than the existing list) but I don't think the `forEach` impl is any less easy to read and it will stop IDEs shouting at us.

Reply:
SGTM

Reply:
TIL and 👍